### PR TITLE
Status Endpoint

### PR DESCRIPTION
**Pull request by movchan74**
## Summary
This update introduces a new status endpoint to the SDK for monitoring status. It also includes additional tests for both the new endpoint and previously implemented functionality.

## Key Changes
1. **Status Endpoint**: Added a new endpoint that returns the SDK status, improving system monitoring capabilities.
2. **Status Endpoint Tests**: Implemented tests to ensure the proper functioning of the new status endpoint.
3. **Deploy Error Tests**: Added missing tests for deploy error handling, originally intended for PR #148.

Closes https://github.com/mobiusml/aana_sdk/issues/162

### CODE REVIEW

File: aana/api/request_handler.py [hunk 1]
```diff
@@ -229,3 +230,46 @@ async def _async_chat_completions(
...
+            app.status == "DEPLOY_FAILED"
+            or app.status == "UNHEALTHY"
+            or app.status == "NOT_STARTED"
+            for app in serve_status.applications.values()
+        ):
+            sdk_status = SDKStatus.UNHEALTHY
+            error_messages = []
+            for app_name, app_status in serve_status.applications.items():
+                if (
+                    app_status.status == "DEPLOY_FAILED"
```
Comment (HRashidi) [inline aana/api/request_handler.py:253]: Should we add the `NOT_STARTED` apps included in message too?

File: aana/api/request_handler.py [hunk 2]
```diff
@@ -229,3 +230,46 @@ async def _async_chat_completions(
...
+        message = ""
+        if any(
+            app.status == "DEPLOY_FAILED"
+            or app.status == "UNHEALTHY"
+            or app.status == "NOT_STARTED"
+            for app in serve_status.applications.values()
+        ):
+            sdk_status = SDKStatus.UNHEALTHY
+            error_messages = []
+            for app_name, app_status in serve_status.applications.items():
```
Comment (HRashidi) [inline aana/api/request_handler.py:251]: Should we check for all the applications on the serve or just the current app?

File: aana/api/request_handler.py [hunk 3]
```diff
@@ -229,3 +230,46 @@ async def _async_chat_completions(
...
+    async def status(self) -> SDKStatusResponse:
+        """The endpoint for checking the status of the application.
+
+        Returns:
+            SDKStatusResponse: The response containing the status of the application.
+        """
+        serve_status = serve.status()
+        message = ""
+        if any(
+            app.status == "DEPLOY_FAILED"
```
Comment (HRashidi) [inline aana/api/request_handler.py:244]: using list is better! 
```app.status in ["DEPLOY_FAILED", ...]```

File: aana/api/request_handler.py [hunk 4]
```diff
@@ -229,3 +230,46 @@ async def _async_chat_completions(
...
+        elif all(app.status == "RUNNING" for app in serve_status.applications.values()):
+            sdk_status = SDKStatus.RUNNING
+        elif any(
+            app.status == "DEPLOYING" or app.status == "DELETING"
+            for app in serve_status.applications.values()
+        ):
+            sdk_status = SDKStatus.DEPLOYING
+        else:
+            sdk_status = SDKStatus.UNHEALTHY
+            message = "Unknown status"
```
Comment (HRashidi) [inline aana/api/request_handler.py:273]: We can return the default message for app(s) here

File: aana/api/request_handler.py (hunk 2 above)
Comment (HRashidi) [reply aana/api/request_handler.py:251]: If yes, maybe we can return the response individually (as a dict) in the response!

File: aana/api/request_handler.py [hunk 5]
```diff
@@ -229,3 +230,46 @@ async def _async_chat_completions(
...
+            sdk_status = SDKStatus.UNHEALTHY
+            error_messages = []
+            for app_name, app_status in serve_status.applications.items():
+                if (
+                    app_status.status == "DEPLOY_FAILED"
+                    or app_status.status == "UNHEALTHY"
+                ):
+                    for (
+                        deployment_name,
+                        deployment_status,
```
Comment (HRashidi) [inline aana/api/request_handler.py:258]: How would be the status of a deployment with more that 1 replica? Will it be "UNHEALTHY" if one of replica failed?

File: aana/api/request_handler.py (hunk 2 above)
Comment (movchan74) [reply aana/api/request_handler.py:251]: First, I want to clarify something. We have Aana apps and Ray Serve apps. They are different. One Aana app consists of multiple Ray Serve apps: one for request hander and one per each Aana deployment. You can run multiple Ray Serve apps and multiple Aana apps on the same Ray cluster. Although it is possible to run multiple Aana apps on the same Ray cluster, right now we mostly have one app per cluster.

The status endpoint should return the status of the current Aana app. Right now, we don't have a way to find which Ray Serve apps on the cluster belong to the current Aana app in the request handler but I will add that. 

So to answer your question, the status endpoint has to check all the Ray Serve apps that belong to it.

We can return detailed information for each Ray Serve app as well. Something like this:

```json
{
  "deployments": {
    "app": {
      "message": "",
      "status": "RUNNING"
    },
    "lowercase_deployment": {
      "message": "Error: Lowercase (lowercase_deployment): A replica's health check failed. This deployment will be UNHEALTHY until the replica recovers or a new deploy happens.",
      "status": "UNHEALTHY"
    },
    "task_queue_deployment": {
      "message": "",
      "status": "RUNNING"
    }
  },
  "message": "",
  "status": "UNHEALTHY"
}
```

Is that what you suggest?

File: aana/api/request_handler.py (hunk 1 above)
Comment (movchan74) [reply aana/api/request_handler.py:253]: What do you mean by "default message of app_status"?

File: aana/api/request_handler.py (hunk 5 above)
Comment (movchan74) [reply aana/api/request_handler.py:258]: Yes, if one of the replicas is unhealthy, the whole app is unhealthy. This is default Ray behaviour.

File: aana/api/request_handler.py (hunk 2 above)
Comment (movchan74) [reply aana/api/request_handler.py]: Added it to the PR. Please, review it again.

File: aana/api/request_handler.py (hunk 1 above)
Comment (HRashidi) [reply aana/api/request_handler.py]: The ray serve status  returns something like this:
```json
applications:
  default:
    status: RUNNING
    message: ''"   <--- this one
    last_deployed_time_s: 1694041157.2211847
    deployments:
      Translator:
        status: HEALTHY
        replica_states:
          RUNNING: 1
        message: ''
      Summarizer:
        status: HEALTHY
        replica_states:
          RUNNING: 1
        message: ''
 ```

**Review by HRashidi — APPROVED** [review]

File: aana/api/request_handler.py (hunk 2 above)
Comment (HRashidi) [reply aana/api/request_handler.py]: Much better

File: aana/api/request_handler.py (hunk 1 above)
Comment (movchan74) [reply aana/api/request_handler.py]: It's empty if there isn't any error.

**Pull request merged**